addons/mrp/models/mrp_production.py

```diff
@@ -742,10 +742,67 @@ def _compute_date_finished(self):
             days_delay = production.bom_id.produce_delay
             date_finished = production.date_start + relativedelta(days=days_delay)
             if production._should_postpone_date_finished(date_finished):
-                workorder_expected_duration = sum(production.workorder_ids.mapped('duration_expected'))
+                workorder_expected_duration = production._compute_expected_duration(production.date_start) or \
+                        sum(production.workorder_ids.mapped('duration_expected'))
                 date_finished = date_finished + relativedelta(minutes=workorder_expected_duration or 60)
             production.date_finished = date_finished
 
+    def _compute_expected_duration(self, start_production):
+        """
+        Compute expected duration of all workorder based on workcenter availability and workorder dependencies
+        Return False if at least one workorder has an unavaible workcenter
+        """
+        if not self.workorder_ids:
+            return 0
+        end_date_per_workcenter = defaultdict(lambda: start_production)
+
+        for wo in self._topological_sort_workorders(self.workorder_ids):
+            date_start = end_date_per_workcenter[wo.workcenter_id]
+            for block_wo in wo.operation_id.blocked_by_operation_ids:
+                date_start = max(date_start, end_date_per_workcenter[block_wo.workcenter_id])
```

Comment:
I would make this in one line and I would use ids as keys in `end_date_per_workcenter`:
```suggestion
            date_start = max(end_date_per_workcenter[wc] for wc in wo.workcenter_id.ids  + wo.operation_id.blocked_by_operation_ids.workcenter_id.ids]])
```